src/scene/procedural.js

```diff
@@ -39,36 +39,31 @@ function calculateNormals(positions, indices) {
     const p1p3 = new Vec3();
     const faceNormal = new Vec3();
 
-    const normals = [];
-
-    // Initialize the normal array to zero
-    for (let i = 0; i < positions.length; i++) {
-        normals[i] = 0;
-    }
+    const normals = new Float32Array(positions.length);
 
     // Accumulate face normals for each vertex
     for (let i = 0; i < triangleCount; i++) {
-        const i1 = indices[i * 3];
-        const i2 = indices[i * 3 + 1];
-        const i3 = indices[i * 3 + 2];
+        const i1 = indices[i * 3    ] * 3;
+        const i2 = indices[i * 3 + 1] * 3;
+        const i3 = indices[i * 3 + 2] * 3;
 
-        p1.set(positions[i1 * 3], positions[i1 * 3 + 1], positions[i1 * 3 + 2]);
-        p2.set(positions[i2 * 3], positions[i2 * 3 + 1], positions[i2 * 3 + 2]);
-        p3.set(positions[i3 * 3], positions[i3 * 3 + 1], positions[i3 * 3 + 2]);
+        p1.set(positions[i1], positions[i1 + 1], positions[i1 + 2]);
```

Comment:
I would not even set these on p1, p2 and p3, but directly evaluate subtraction of them to p1p2 and p1p2, something along
`p1p2.x = positions[i2] - positions[i1];